src/ips.rs

```diff
@@ -0,0 +1,525 @@
+        // location. Now we can compute each node's constant rating based on some graph params.
+        // Vector contains IpAddr, node index (from the state.nodes) and rating. We need index just
+        // to be able to easily get the node from nodes vector after sorting.
+        let mut const_factors = Vec::with_capacity(state.nodes.len());
+        for node_idx in 0..state.nodes.len() {
+            let ip = IpAddr::from_str(state.nodes[node_idx].ip.as_str()).unwrap();
+            const_factors.push((
+                ip,
+                node_idx,
+                self.rate_node(
+                    &state.nodes[node_idx],
+                    *degrees.get(&ip).unwrap(), // should be safe to unwrap here
+                    *eigenvalues.get(&ip).unwrap(),
+                ),
+            ));
+        }
+
+        // Iterate over nodes to generate peerlist entry for each node
+        for (node_idx, node) in state.nodes.iter().enumerate() {
+            let node_ip = IpAddr::from_str(node.ip.as_str()).unwrap();
+
+            // Clone const factors for each node to be able to modify them
+            let mut peer_ratings = const_factors.clone();
+            let mut curr_peer_ratings: Vec<(IpAddr, usize, f64)> = Vec::new();
+
+            let mut peer_list_entry = Peer {
+                ip: node_ip,
+                list: Vec::new(),
+            };
+
+            // 1 - update ranks by location for specified node
+            // This need to be done every time as location ranking will change for differently
+            // located nodes.
+            if self.config.use_geolocation {
+                self.update_rating_by_location(node, &state.nodes, &mut peer_ratings);
+            }
+
+            // Load peerlist with current connections (we don't want to change everything)
+            for peer in &agraph[node_idx] {
+                peer_list_entry
+                    .list
+                    .push(IpAddr::from_str(state.nodes[*peer].ip.as_str()).unwrap());
+
+                // Remeber current peer ratings
+                curr_peer_ratings.push(peer_ratings[*peer]);
+            }
+
+            // 2 - Calculate desired vertex degree
+            // In the first iteration we will use avg of degree and degree_delta so all
+            // nodes should pursue to degree_delta level. That could be bad if graph's vertexes
+            // have very high (or low) degrees and therefore, delta is very high (or low) too. But until
+            // we have some better idea this one is the best we can do to keep up with the graph.
+            let desired_degree = (degree_avg as u32 + degrees.get(&node_ip).unwrap()) / 2;
+
+            // 3 - Calculate how many peers to add or delete from peerlist
+            let peers_to_delete_count = if desired_degree < *degrees.get(&node_ip).unwrap() {
+                degrees.get(&node_ip).unwrap() - desired_degree
```

Comment:
Is there any guarantee this won't overflow?

Reply:
The logic would prevent from overflow (underflow) here but added saturated commands just to prevent it in case if at any point in the future the logic would change.